pilot/pkg/proxy/envoy/v2/discovery.go

```diff
@@ -79,6 +78,40 @@ type DiscoveryServer struct {
 	// DebugConfigs controls saving snapshots of configs for /debug/adsz.
 	// Defaults to false, can be enabled with PILOT_DEBUG_ADSZ_CONFIG=1
 	DebugConfigs bool
+
+	// mutex protecting global structs updated or read by ADS service, including EDSUpdates and
+	// shards.
+	mutex sync.RWMutex
+
+	// ServiceShards for a service. This is a global (per-server) list, built from
+	// incremental updates.
+	EndpointShardsByService map[string]*model.ServiceShards
```

Comment:
ServiceInstanceIndicesByService

Reply:
I hate 'ServiceInstance' - what does it mean ? Is it a service or an instance ? And the intent is to keep shards - not the full list of instances. K8S uses Service and Endpoint, Envoy uses Cluster and Endpoint. Why would we call the endpoints 'ServiceInstances' ?

Reply:
Well, if you had named EndpointShardsByService in a clearer way, we wouldn't be arguing.. The comment says ServiceShards for a Service.. The variable is called EndpointShardsByService. Then there is also EndpointShard which is another grouping of endpoints.. Everything is either a shard or shards or xByService.. Essentially what Kuat did two years ago and why you don't understand the code today. And the pattern will repeat

Reply:
Happy with EndpointShardsByService :-)